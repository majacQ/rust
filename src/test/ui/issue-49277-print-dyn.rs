// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
// file at the top-level directory of this distribution and at
// http://rust-lang.org/COPYRIGHT.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

use std::fmt::{Debug, Display};

macro_rules! demonstrate_type_mismatch {
    ($name:ident, $value:expr, $type1:ty, $type2:ty) => {
        fn $name() {
            let x: $type1 = $value;
            let _: $type2 = x;
        }
    }
}

macro_rules! demonstrate_trait_object_type_mismatches {
    ($value:expr, $T:path, $U:path) => {
        demonstrate_type_mismatch!(_box, Box::new($value), Box<$T>, Box<$U>);
        demonstrate_type_mismatch!(_ref, &$value, &dyn $T, &dyn $U);
        demonstrate_type_mismatch!(_mut_ref, &mut $value, &mut dyn $T, &mut dyn $U);
        demonstrate_type_mismatch!(_slice_ref, &[&$value], &[&dyn $T], &[&dyn $U]);
    }
}

demonstrate_trait_object_type_mismatches!(1, Debug, Display);

fn main() {}
