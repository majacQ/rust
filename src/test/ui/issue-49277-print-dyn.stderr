error[E0308]: mismatched types
  --> $DIR/issue-49277-print-dyn.rs:17:29
   |
LL |             let _: $type2 = x;
   |                             ^ expected trait `std::fmt::Display`, found trait `std::fmt::Debug`
...
LL | demonstrate_trait_object_type_mismatches!(1, Debug, Display);
   | ------------------------------------------------------------- in this macro invocation
   |
   = note: expected type `std::boxed::Box<dyn std::fmt::Display>`
              found type `std::boxed::Box<dyn std::fmt::Debug>`

error[E0308]: mismatched types
  --> $DIR/issue-49277-print-dyn.rs:17:29
   |
LL |             let _: $type2 = x;
   |                             ^ expected trait `std::fmt::Display`, found trait `std::fmt::Debug`
...
LL | demonstrate_trait_object_type_mismatches!(1, Debug, Display);
   | ------------------------------------------------------------- in this macro invocation
   |
   = note: expected type `&dyn std::fmt::Display`
              found type `&dyn std::fmt::Debug`

error[E0308]: mismatched types
  --> $DIR/issue-49277-print-dyn.rs:17:29
   |
LL |             let _: $type2 = x;
   |                             ^ expected trait `std::fmt::Display`, found trait `std::fmt::Debug`
...
LL | demonstrate_trait_object_type_mismatches!(1, Debug, Display);
   | ------------------------------------------------------------- in this macro invocation
   |
   = note: expected type `&mut dyn std::fmt::Display`
              found type `&mut dyn std::fmt::Debug`

error[E0308]: mismatched types
  --> $DIR/issue-49277-print-dyn.rs:17:29
   |
LL |             let _: $type2 = x;
   |                             ^ expected trait `std::fmt::Display`, found trait `std::fmt::Debug`
...
LL | demonstrate_trait_object_type_mismatches!(1, Debug, Display);
   | ------------------------------------------------------------- in this macro invocation
   |
   = note: expected type `&[&dyn std::fmt::Display]`
              found type `&[&dyn std::fmt::Debug]`

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0308`.
